Migrating from dockershim documentation

SergeyKanzhelev opened the pull request:
/sig node

Fixes: https://github.com/kubernetes/website/issues/25768

This is the first iteration of a migration from dockershim instructions. PR is marked as WIP to collect initial feedback on a structure and content.

KEP: https://github.com/kubernetes/enhancements/issues/2221

sftim commented:
:information_source:  Watch out - you've a space in a path.

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/ migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,99 @@
+
+<!-- overview -->
+
+Docker is being deprecated. From the [Dockershim Deprecation
+FAQ](https://kubernetes.io/blog/2020/12/02/dockershim-faq/) you might already
+know that most apps do not have a direct dependency on runtime hosting
+containers. However, there are still a lot of telemetry and security agents
+that has a dependency on docker to collect containers metadata, logs and
+metrics. This document aggregates information on how to detect these
+dependencies and links on how to migrate these agents to use generic tools or
+alternative runtimes.
```
To help pages stand the test of time, try to write wording that works both during and after the deprecation. It's not always possible but when it can be done it avoids work to track and reword things after the fact.

I've seen quite a few dated pages on the Kubernetes docs so I try to word assuming that the cleanup probably gets forgotten.

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/ migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,99 @@
+
+## Telemetry and security agents
+
+There are a few ways agents may be run on kubernetes cluster. Agents have a direct dependency on Docker when they run on nodes or as daemonsets. Some telemetry and security agents may rely on certain logs or metrics specific for Docker that would not be available any longer. This document does not cover these cases.
```
nit: “Kubernetes” is a proper noun; we write object kinds like DaemonSet in UpperCamelCase.

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/ migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,99 @@
+
+Historically, kubernetes was built on top of Docker. Kubernetes is managing networking and scheduling, Docker was placing and operating containers on a node. So you can get scheduling-related metadata like a pod name from Kubernetes and containers state information from Docker. Over time more runtimes were created to manage containers. Also there are projects and kubernetes features that generali ...
+
+But some agents are still Docker-dependent. They may run commands like [docker
+ps](https://docs.docker.com/engine/reference/commandline/ps/) or [docker
```
```suggestion
Some agents are tied specifically to the Docker tool. The agents may run commands like [`docker
 ps`](https://docs.docker.com/engine/reference/commandline/ps/) or [`docker
```
(etc)
?

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/ migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,99 @@
+commands will not work any longer.
+
+
+### Detecting Docker dependency from the agents-daemonsets
```
This is a task page, so I suggest
```suggestion
### Identify DaemonSets that depend on Docker {#identify-docker-dependency}
```

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/ migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,99 @@
+
+### Detecting Docker dependency from the agents-daemonsets
+
+If a pod wants to make calls to the `dockerd` running on the node, the pod must mount the docker daemon as a volume.
```
```suggestion
If a pod wants to make calls to the `dockerd` running on the node, the pod must either:
- mount the filesystem containing the Docker daemon's privileged socket, as a {{< glossary_tooltip text="volume" term_id="volume" >}}; or
- mount the specific path of the Docker daemon's privileged socket directly, also as a volume.
```
?

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/ migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,99 @@
+
+If a pod wants to make calls to the `dockerd` running on the node, the pod must mount the docker daemon as a volume.
+
+For example, On COS images, docker exposes it’s unix socket at
```
nits:
```suggestion
For example: on COS images, Docker exposes its Unix domain socket at
```

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/ migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,99 @@
+If a pod wants to make calls to the `dockerd` running on the node, the pod must mount the docker daemon as a volume.
+
+For example, On COS images, docker exposes it’s unix socket at
+`/var/run/docker.sock` This means that the customer’s pod spec will include a
```
can we assume that there is a customer?

SergeyKanzhelev replied:
fixed

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/ migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,99 @@
+`/var/run/docker.sock` This means that the customer’s pod spec will include a
+`hostPath` volume mount of `/var/run/docker.sock`.
+
+The script to find all pods mounting this volume may look like this:
```
```suggestion
Here's a sample shell script to find Pods that have a mount directly
mapping the Docker socket:
```

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/ migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,99 @@
+
+(this script outputs the namespace and name of the pod. You can remove the grep `/var/run/docker.sock` to review other mounts)
+
+Note, there are alternative ways to call a Docker. For instance, the parent
+directory `/var/run` may be mounted instead of the full path (like in [this
+example](https://gist.github.com/itaysk/7bc3e56d69c4d72a549286d98fd557dd)). So
+the script above will only detect the most common uses.
```
```suggestion
{{< note >}}
There are alternative ways for a pod to access Docker on the host. For instance, the parent
directory `/var/run` may be mounted instead of the full path (like in [this
example](https://gist.github.com/itaysk/7bc3e56d69c4d72a549286d98fd557dd)).
The script above only detects the most common uses.
{{< /note >}}
```

Is this acceptable for a task page? If we only explain some of the easier cases, readers might feel short changed.

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/ migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,99 @@
+In case your cluster nodes are customized and install additional security and
+telemetry agents on the node, make sure to check with the vendor of the agent whether it has dependency on Docker.
+
+## Telemetry and security agent vendors
+
+This section is intended to collect information about various telemetry and
+security agents that may have dependency on containers runtime.
+
+This section is work in progress.
+
+{{< note >}}
+
+Vendors, please help to fill out this section. See
+[CONTRIBUTING](https://github.com/kubernetes/website/blob/master/CONTRIBUTING.md)
+guide for instructions. Or create an issue if you need help formatting the
+document.
+
+{{< /note >}}
```
I would omit this, replacing it with a draft pull request once the other documentation has landed. We don't publish work-in-progress documentation.

Alternatively, hyperlink to a wiki page run by SIG Node, or something similar. Linking to incomplete content is OK.

SergeyKanzhelev replied:
Would the list of vendors who likely has dependency on Docker be OK to have alongside with this request? Let me draft it...

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/ migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,99 @@
+
+{{< /note >}}
+
+### SignalFx (Splunk)
```
We prefer to hyperlink to vendor docs (see https://kubernetes.io/docs/contribute/style/content-guide/ for policy). Is that feasible? I think if we let vendors know that this is our policy, vendors ought to be quite willing to add the documentation and to let us know its URL.

If a vendor does not wish to publish documentation, that is fine, we can omit the link.

SergeyKanzhelev replied:
@jrcamp what do you think?

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/find-if-you-affected.md:
```diff
@@ -0,0 +1,62 @@
+---
+title: Finding if you are affected by dockershim deprecation
```
```suggestion
title: Check whether Dockershim deprecation affects you
```

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/find-if-you-affected.md:
```diff
@@ -0,0 +1,62 @@
+
+<!-- overview -->
+
+This is very unlikely that your application has a dependency on dockershim. In
+order to understand how your cluster or apps may acquire dependency on Docker as
+a container runtime, you need to understand what role dockershim plays when
+container is being scheduled on the node.
```
I suggest:
```suggestion
{{< feature-state for_k8s_version="v1.20" state="deprecated" >}}

The `dockershim` component of Kubernetes implements a
{{< glossary_tooltip term_id="cri" text="Container Runtime Interface">}} (CRI) integration with Docker.
Kubernetes' built-in `dockershim` component was deprecated in release v1.20.

This page explains how your cluster could be using Docker as a
{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}},
provides details on the role that `dockershim` plays when in use, and shows steps
you can take to check whether any workloads could be affected.
```

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/find-if-you-affected.md:
```diff
@@ -0,0 +1,62 @@
+
+## Container runtimes
+
+[Container runtime](/docs/setup/production-environment/container-runtimes/) is a software that can schedule Pods. Kubernetes is responsible for orchestration of application Pods, and uses CRI (container runtime interface) to communicate to container runtime to actually schedule these Pods and get their status.
```
```suggestion
A [container runtime](/docs/concepts/containers/#container-runtimes) is software that can
execute the containers that make up a Kubernetes pod. Kubernetes is responsible for orchestration
and scheduling of Pods; on each node, the {{< glossary_tooltip text="kubelet" term_id="kubelet" >}}
uses the container runtime interface as an abstraction so that you can use any compatible
container runtime.
```

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/find-if-you-affected.md:
```diff
@@ -0,0 +1,62 @@
+
+[Container runtime](/docs/setup/production-environment/container-runtimes/) is a software that can schedule Pods. Kubernetes is responsible for orchestration of application Pods, and uses CRI (container runtime interface) to communicate to container runtime to actually schedule these Pods and get their status.
+
+There are various container runtimes, each has special features and decide on tradeoffs they make between performance, security and functionality. Dockershim is a first container runtime used by Kubernetes. When dockershim is used as a container runtime, kubelet communicates with Docker to schedule containers and check on it’s status. Docker in its turn uses Containerd to actually schedule contai ...
```
```suggestion
In its earliest releases, Kubernetes offered compatibility with just one container runtime: Docker.
Later in the Kubernetes project's history, cluster operators wanted to adopt additional container runtimes.
The CRI was designed to allow this kind of flexibility - and the kubelet began supporting CRI. However,
because Docker existed before the CRI specification was invented, the Kubernetes project created an
adapter component, “dockershim”. The dockershim adapter allows the kubelet to interact with Docker as
if Docker were a CRI compatible runtime.
```

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/find-if-you-affected.md:
```diff
@@ -0,0 +1,62 @@
+
+You can read about it in [kubernetes containerd integration goes GA](/blog/2018/05/24/kubernetes-containerd-integration-goes-ga/) blog post.
+
+![Dockershim vs. CRI with Containerd](/images/blog/2018-05-24-kubernetes-containerd-integration-goes-ga/cri-containerd.png)
+
+Switching to containerd as a container runtime eliminates the middleman. All the
+same containers can be run by container runtimes like Containerd as before. But
+now, since containers were scheduled directly with the container runtime, they
+are not visible to Docker. So any Docker tooling or fancy UI you might have used
+before to check on these containers is no longer available.
+
+You cannot get container information using `docker ps` or `docker inspect`
+commands. As you cannot list containers, you cannot get logs, stop containers,
+or execute something with `docker exec`.
+
+{{< note >}}
+
+Stopping containers via Docker CLI was never a good idea as it is
+better to allow Kubernetes to orchestrate them. So you don’t confuse Kubernetes
+by disappearing containers.
+
+{{< /note >}}
+
+You can still pull images or build them using `docker build` command. But images
+built or pulled by Docker would not be visible to container runtime and
+Kubernetes. They needed to be pushed to some registry to allow them to be used
+by kubernetes.
```
I'd either move this to the end of the task, or omit it. Readers are likely to be keen to get to the meat of the page; that is, “Finding if your app has a dependency on Docker”.

SergeyKanzhelev replied:
moved to the end and rephrased. Good suggestion

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/find-if-you-affected.md:
```diff
@@ -0,0 +1,62 @@
+   - Monitoring and security agents installed on nodes directly
+3. Third-party tools that perform above mentioned privileged operations. See
+   [Migrating telemetry and security agents from
+   dockershim](migrating-telemetry-and-security-agents.md) for more information.
```
nit: broken hyperlink

SergeyKanzhelev replied:
do you mean `.md`? Always forgetting. Sorry if it is being annoying.

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/_index.md:
```diff
@@ -0,0 +1,39 @@
+
+## {{% heading "whatsnext" %}}
+
+{{< note >}}
+
+The following sections may be added in future:
+
+- Dockershim deprecation and Windows
+- Known issues of non-dockershim runtimes
+- Transition from docker CLI to crictl
+- Building images on kubernetes without Docker
+
+{{< /note >}}
```
We usually [avoid statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future) in published documentation. It's OK to say “foo _has been_ deprecated” but we prefer not to hint “we might document this more some day”.

SergeyKanzhelev replied:
I will remove this before merging this PR

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/_index.md:
```diff
@@ -0,0 +1,39 @@
+installations. You can find these links useful to understand the problem better:
+
+- [Don't Panic: Kubernetes and Docker](/blog/2020/12/02/dont-panic-kubernetes-and-docker/)
+- [Dockershim Deprecation FAQ](https://kubernetes.io/blog/2020/12/02/dockershim-faq/)
```
nit:
```suggestion
- [Dockershim Deprecation FAQ](/blog/2020/12/02/dockershim-faq/)
- ```

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/_index.md:
```diff
@@ -0,0 +1,39 @@
+- [Don't Panic: Kubernetes and Docker](/blog/2020/12/02/dont-panic-kubernetes-and-docker/)
+- [Dockershim Deprecation FAQ](https://kubernetes.io/blog/2020/12/02/dockershim-faq/)
+
+We highly encourage you to consider various migration options. Check out
```
If you can, [avoid using “we”](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-using-we).

SergeyKanzhelev replied:
fixed. thx

sftim reviewed (commented):
Thanks @SergeyKanzhelev 

I've added in a bunch of feedback. I appreciate that this is all WIP so please treat these as _hints_ and not as instructions that you must follow.

SergeyKanzhelev commented:
I addressed most of the feedback. I will add more content that I want to have in a first iteration and ping you again. Likely on Monday.

Thank you for your thoughtful review!

tengqm commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/_index.md:
```diff
@@ -0,0 +1,39 @@
+with the migration. So the issue can be fixed timely and your cluster would be
+ready for dockershim removal.
+
+[dockershim deprecation]: /blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation
```
I'd suggest we avoid this style of markdown links because it is weird syntax for the link checker.
Please consider use inline links if you have another iteration.

SergeyKanzhelev replied:
done

tengqm commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you.md:
```diff
@@ -0,0 +1,85 @@
+   - Monitoring and security agents installed on nodes directly.
+3. Third-party tools that perform above mentioned privileged operations. See
+   [Migrating telemetry and security agents from
+   dockershim](migrating-telemetry-and-security-agents) for more information.
```
please try avoid relative links ... the convention here is always have the English intrasite links start with "/docs"

tengqm commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you.md:
```diff
@@ -0,0 +1,85 @@
+3. Third-party tools that perform above mentioned privileged operations. See
+   [Migrating telemetry and security agents from
+   dockershim](migrating-telemetry-and-security-agents) for more information.
+4. Make sure there is no indirect dependencies on dockershim behavior. Some
```
This sounds pretty challenging for a reader who knows nothing about kubelet code.

kbhawkey replied:
Make sure there are no indirect dependencies? How is this done?

SergeyKanzhelev replied:
expanded

tengqm commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you.md:
```diff
@@ -0,0 +1,85 @@
+![Dockershim vs. CRI with Containerd](/images/blog/2018-05-24-kubernetes-containerd-integration-goes-ga/cri-containerd.png)
+
+Switching to containerd as a container runtime eliminates the middleman. All the
+same containers can be run by container runtimes like Containerd as before. But
```
Please try be consistent when spelling "containerd".

tengqm commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,118 @@
+
+Historically, kubernetes was built on top of Docker. Kubernetes is managing networking and scheduling, Docker was placing and operating containers on a node. So you can get scheduling-related metadata like a pod name from Kubernetes and containers state information from Docker. Over time more runtimes were created to manage containers. Also there are projects and kubernetes features that generali ...
+
+Some agents are tied specifically to the Docker tool. The agents may run commands like [`docker
+ ps`](https://docs.docker.com/engine/reference/commandline/ps/) or [`docker
+top](https://docs.docker.com/engine/reference/commandline/top/) to list
+containers and processes or [docker
+logs](https://docs.docker.com/engine/reference/commandline/logs/) to subscribe
```
The wrapping of lines here are terrible. It would be desirable to have each link stay on the same line, no matter how long that line may be.

There is a missing backtiq for the docker top command.

SergeyKanzhelev replied:
> The wrapping of lines here are terrible.

agree. This extension I use always doing it (https://github.com/stkb/Rewrap/issues/112). 

Fixed the backtiq

tengqm commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,118 @@
+The script above only detects the most common uses.
+{{< /note >}}
+
+### Detecting Docker dependency from node deployed agents
```
```suggestion
### Detecting Docker dependency from node agents
```
?

tengqm commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,118 @@
+This section is intended to collect information about various telemetry and
+security agents that may have dependency on containers runtime.
+
+### SignalFx (Splunk)
```
I'm a little bit struggled at the section.
The overall guideline is that we are trying to avoid dual-hosted contents. However, it seems to me like we don't have a lot of options as of now ...

SergeyKanzhelev replied:
I know. We can start with this and see whether vendors would love their own pages with instructions.

https://github.com/kubernetes/website/pull/25787#discussion_r548292025

SergeyKanzhelev commented:
Thank you @tengqm and @sftim . I think we can merge it as it is now and I will be working on other pages I wanted later. Can you please make a final review for this part.




Ideas for other tasks in this section:

- migrating Windows workloads from Dockershim
- work around the known issues of non-dockershim runtimes
- transition from docker CLI to crictl
- building images on kubernetes without Dockershim

neolit123 reviewed (commented):
thanks for the write up @SergeyKanzhelev 
as i mentioned on the original PR that did the deprecation, i think something important that we need to document is how the users can actually do an in-place swap of the container runtime on node.

SergeyKanzhelev commented:
> as i mentioned on the original PR that did the deprecation, i think something important that we need to document is how the users can actually do an in-place swap of the container runtime on node.

In-place may be one of the strategies. And vendors/admins would need to configure nodes to enable this. Simple node recreation and drain may be sufficient enough for many users. Similar to how regular upgrade to another k8s version works.

That's said, this PR creates a good place for this task description.

neolit123 commented:
> In-place may be one of the strategies. And vendors/admins would need to configure nodes to enable this. Simple node recreation and drain may be sufficient enough for many users. Similar to how regular upgrade to another k8s version works.

re-place upgrades are what most hosted vendor solutions are doing.
as part of that process, the vendors could introduce new nodes that have a CR different than Docker.

but bare metal, self-hosted hardware users do in-place upgrades, by draining and upgrading node kubelets.
similar steps would apply if they wish to replace the container runtime.
these users don't have the option to introduce new nodes and ideally we should document the in-place steps for them.

> That's said, this PR creates a good place for this task description.

do you have plans to document this as part of this PR?

SergeyKanzhelev commented:
> do you have plans to document this as part of this PR?

Not in this PR. I have more topics to cover as mentioned in previous comment as well:

- migrating Windows workloads from Dockershim
- work around the known issues of non-dockershim runtimes
- transition from docker CLI to crictl
- building images on kubernetes without Dockershim

Also I'm not super familiar with how the bare metal in-place upgrades typically work. Do you have any example documentation that I can use for reference? Or maybe we can find somebody knowing these scenarios well to document them.

I cannot tell whether this document is critically important. But if it is, maybe you can send a update for KEP to make it a Step 2 graduation criteria?

neolit123 commented:
> > do you have plans to document this as part of this PR?
> 
> Not in this PR. I have more topics to cover as mentioned in previous comment as well:
> 
> * migrating Windows workloads from Dockershim
> * work around the known issues of non-dockershim runtimes
> * transition from docker CLI to crictl
> * building images on kubernetes without Dockershim
> 
> Also I'm not super familiar with how the bare metal in-place upgrades typically work. Do you have any example documentation that I can use for reference? Or maybe we can find somebody knowing these scenarios well to document them.
> 

kubeadm's default upgrade command performs an in-place upgrade:
https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/

the process would roughly be:
- drain a node
- stop the docker runtime
- start new runtime
- change the kubelet socket
- restart kubelet
- uncordon the node

except that i don't know if that is completely valid and supported.
if it's not supported our documentation needs to tell that to the users, which 

> I cannot tell whether this document is critically important. But if it is, maybe you can send a update for KEP to make it a Step 2 graduation criteria?

i'd estimate it as more important than these topics:

> - migrating Windows workloads from Dockershim
> - work around the known issues of non-dockershim runtimes
> - transition from docker CLI to crictl
> - building images on kubernetes without Dockershim

in fact i wanted it documented before we merged the deprecation KEP:
https://github.com/kubernetes/kubernetes/pull/94624#issuecomment-689562610

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/_index.md:
```diff
@@ -0,0 +1,23 @@
+---
+title: "Migrating from dockershim to other container runtimes"
+weight: 10
+---
```
NB I recommend setting a `content_type`

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you.md:
```diff
@@ -0,0 +1,85 @@
+
+{{< note >}}
+
+Stopping containers via Docker CLI was never a good idea as it is
+better to allow Kubernetes to orchestrate them. So you don’t confuse Kubernetes
+by disappearing containers.
```
```suggestion
If you're running workloads via Kubernetes, the best way to stop a container is through
the Kubernetes API rather than directly through the container runtime (this advice applies
for all container runtimes, not just Docker).
```
?

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you.md:
```diff
@@ -0,0 +1,85 @@
+
+{{< /note >}}
+
+You can still pull images or build them using `docker build` command. But images
+built or pulled by Docker would not be visible to container runtime and
+Kubernetes. They needed to be pushed to some registry to allow them to be used
+by kubernetes.
```
I'm not sure who is the audience for this paragraph. Is it people who are running `docker build` directly on
a node?

SergeyKanzhelev replied:
yes. It is not uncommon to create DIY continuous integration tasks or such by running `docker build` on nodes. Linking alternative ways to build containers like https://github.com/GoogleContainerTools/kaniko or [docker buildx suggested in AKS](https://github.com/Azure/AKS/releases/tag/2020-09-21#tag-select-menu-844f6e74-4aee-11eb-80e5-5b2c23eaefc6:~:text=.%20If%20you%20are%20doing%20container%20builds%20in%20cluster%20please%20use%20the%20recommended%20docker%20buildx.) would be very helpful.  And this is one of the next tasks I want to add.

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,121 @@
+This section is intended to collect information about various telemetry and
+security agents that may have dependency on containers runtime.
+
+### SignalFx (Splunk)
```
```suggestion
### SignalFx (Splunk) {#vendor-signalfx-splunk}
```
@SergeyKanzhelev this seems to be outside what the [content guide](https://kubernetes.io/docs/contribute/style/content-guide/) permits. I'm going to
/hold
this PR for now; we can unhold it if it becomes clear that these vendor-specific details are OK to add.

SergeyKanzhelev replied:
agree! As I said, one blocker for some user is to transition monitoring agents to the new runtime. So it will be super helpful to either accumulate this information here or has a directory of links for easy discoverability.

sftim reviewed (changes requested):
@SergeyKanzhelev I'm afraid I do have some strong concerns that this is not yet OK to merge.

At the time of my review this PR is not marked as WIP, and proposes a change that seems to contravene the content guide on third party content. I've added a hold to make that concern clear.

Additionally, https://github.com/kubernetes/website/pull/25787#discussion_r548332121 suggests that the PR should in fact still me marked as WIP.

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,121 @@
+The pod accessing Docker may be named like:
+- `signalfx-agent`
+
+### Other agents
+
+We keep the work in progress version of migration instructions in
+[Google doc](https://docs.google.com/document/d/1ZFi4uKit63ga5sxEiZblfb-c23lFhvy6RXVPikS8wf0/edit#).
+
+{{< note >}}
+
+Vendors, please help to fill out this section. See
+[CONTRIBUTING](https://github.com/kubernetes/website/blob/master/CONTRIBUTING.md)
+guide for instructions. Or create an issue if you need help formatting the
+document.
+
+{{< /note >}}
```
(this section is not complete; also, I don't think we'd accept any of those PRs)

SergeyKanzhelev replied:
Do you have any thoughts on what would be the appropriate number of vendor docs would be for the initial version of the document? Is there any prior art in this repo we can learn from?

sftim replied:
> the appropriate number of vendor docs 

Given https://github.com/kubernetes/enhancements/tree/master/keps/sig-docs/1326-third-party-content-in-docs#readme, zero. We can link to vendor documentation for this sort of thing but we don't host it.

SergeyKanzhelev replied:
I removed vendor docs and kept the link to the google doc

SergeyKanzhelev commented:
> kubeadm's default upgrade command performs an in-place upgrade:
> https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/
> ...
> except that i don't know if that is completely valid and supported.

@neolit123 so the upgrade docs recommend to fully drain the node, reinstall kubelet and start it back. What would you expect can go wrong here if instead of reinstall of kubelet, there will be a reconfiguration step?

Also, do you think we should add the change of runtime step into the [upgrade doc](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/) instead? Since dockershim is deprecated, suggestion of a runtime change on upgrade may be the most suitable.

In any case, creating an issue to track this: https://github.com/kubernetes/website/issues/25879

SergeyKanzhelev commented:
@sftim, if the resolution of the question of incomplete documentation will take a while, I can replace it with the link to google doc and merge this as initial version. Would this be acceptable?

neolit123 commented:
> @neolit123 so the upgrade docs recommend to fully drain the node, reinstall kubelet and start it back. What would you expect can go wrong here if instead of reinstall of kubelet, there will be a reconfiguration step?

that was my question as well.

sig node did not comment on this previously:
https://github.com/kubernetes/kubernetes/pull/94624#issuecomment-690729820

> Also, do you think we should add the change of runtime step into the upgrade doc instead? Since dockershim is deprecated, suggestion of a runtime change on upgrade may be the most suitable.

depends on how complex the step-by-step would be.
we could cross link the kubeadm upgrade guide to a new page about this.

kbhawkey commented:
/cc @kubernetes/sig-node-pr-reviews

kbhawkey commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you.md:
```diff
@@ -0,0 +1,85 @@
+
+<!-- overview -->
+
+{{< feature-state for_k8s_version="v1.20" state="deprecated" >}}
```
Does the feature state shortcode apply in this case?

SergeyKanzhelev replied:
@sftim can you suggest?

sftim replied:
I wouldn't use the shortcode here.

kbhawkey commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/_index.md:
```diff
@@ -0,0 +1,24 @@
+---
+title: "Migrating from dockershim to other container runtimes"
```
nit: Could shorten this title
Migrating from dockershim

SergeyKanzhelev replied:
done

kbhawkey commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/_index.md:
```diff
@@ -0,0 +1,24 @@
+Since the announcement of [dockershim deprecation](/blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation)
+in Kubernetes 1.20, there were questions on how this will affect various workloads and Kubernetes
+installations. You can find these links useful to understand the problem better:
+
```
nit: Is there one blog that refers to the other blogs? I think one link could bring in the other articles.

SergeyKanzhelev replied:
agree. Reworded and only left the FAQ

kbhawkey commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you.md:
```diff
@@ -0,0 +1,85 @@
+---
+title: Check whether Dockershim deprecation affects you
```
Is the focus of this page to:
- help cluster administrators determine if their cluster is using Docker and `dockershim`?
- help cluster administrators add/configure another container runtime?
- provide steps to update running applications that may be tightly integrated with Docker?

SergeyKanzhelev replied:
Intention was this:

> - provide steps to update running applications that may be tightly integrated with Docker?

Good question, perhaps this would mean the better place for this doc would be one level up (not under cluster administration)

kbhawkey commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you.md:
```diff
@@ -0,0 +1,85 @@
+Kubernetes' built-in `dockershim` component was deprecated in release v1.20.
+
+This page explains how your cluster could be using Docker as a
+{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}},
```
nit: The container runtime glossary tooltip is used twice.

You could add links to the sections that follow (and create custom anchor text to link to). Something like:
```
This page [explains the role of `dockershim`](#role-of-dockershim) and provides
 [instructions to check for `dockershim` dependencies](#find-docker-dependencies).
```

kbhawkey commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you.md:
```diff
@@ -0,0 +1,85 @@
+provides details on the role that `dockershim` plays when in use, and shows steps
+you can take to check whether any workloads could be affected by `dockershim` deprecation.
+
+## Finding if your app has a dependencies on Docker
```
Add custom anchor, something like:

```
## Finding if your app has a dependencies on Docker {#find-docker-dependencies}
```

SergeyKanzhelev replied:
done

kbhawkey commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you.md:
```diff
@@ -0,0 +1,85 @@
+as a dependency on Docker as a container runtime.
+
+When alternative container runtime is used, executing Docker commands may either
+not work or yeild unexpected output. This is how you can find whether you have a
```
nit: spelling, s/yeild/yield/

kbhawkey commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you.md:
```diff
@@ -0,0 +1,85 @@
+not work or yeild unexpected output. This is how you can find whether you have a
+dependency on Docker:
+
+1. Make sure no privileged Pods executing Docker commands.
```
Make sure no privileged Pods execute Docker commands.

kbhawkey commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you.md:
```diff
@@ -0,0 +1,85 @@
+
+1. Make sure no privileged Pods executing Docker commands.
+2. Check that scripts and apps running on nodes outside of Kubernetes
+   infrastructure do not execute docker commands. It might be:
```
it: ... do not execute Docker commands

kbhawkey commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you.md:
```diff
@@ -0,0 +1,85 @@
+   tooling may be configured to react to Docker-specific behaviors, for example
+   specific logs or metrics. This is an edge case.
+
+## Dependency on Docker explained
```
nit: Could add (change as needed):
```
{#role-of-dockershim}
```

SergeyKanzhelev replied:
this was moved after as this is the task page. So we want to prioritize steps over the explanation.

kbhawkey commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you.md:
```diff
@@ -0,0 +1,85 @@
+Later in the Kubernetes project's history, cluster operators wanted to adopt additional container runtimes.
+The CRI was designed to allow this kind of flexibility - and the kubelet began supporting CRI. However,
+because Docker existed before the CRI specification was invented, the Kubernetes project created an
+adapter component, “dockershim”. The dockershim adapter allows the kubelet to interact with Docker as
```
nit: use backticks instead of double quotes around dockershim?
`dockershim`

kbhawkey commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you.md:
```diff
@@ -0,0 +1,85 @@
+
+Switching to Containerd as a container runtime eliminates the middleman. All the
+same containers can be run by container runtimes like Containerd as before. But
+now, since containers were scheduled directly with the container runtime, they
```
possible edit (change as needed):
But now, since containers schedule directly with the container runtime, they are not visible to Docker.

kbhawkey commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you.md:
```diff
@@ -0,0 +1,85 @@
+
+You cannot get container information using `docker ps` or `docker inspect`
+commands. As you cannot list containers, you cannot get logs, stop containers,
+or execute something with `docker exec`.
```
nit: ... or execute commands such as `docker exec`.

SergeyKanzhelev replied:
clarified in text. Docker exec is used to execute someting inside the container.

kbhawkey commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you.md:
```diff
@@ -0,0 +1,85 @@
+You can still pull images or build them using `docker build` command. But images
+built or pulled by Docker would not be visible to container runtime and
+Kubernetes. They needed to be pushed to some registry to allow them to be used
+by kubernetes.
```
You can still pull images or build images using `docker build`. However, images built or pulled by Docker are not visible to the container runtime and Kubernetes. You must push container images to a registry to allow them to be used by Kubernetes.

s/kubernetes/Kubernetes/

kbhawkey commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,121 @@
+containers. However, there are still a lot of telemetry and security agents
+that has a dependency on docker to collect containers metadata, logs and
+metrics. This document aggregates information on how to detect these
+dependencies and links on how to migrate these agents to use generic tools or
```
nit: Could remove: 
"However, there are still a lot of telemetry and security agents ..."

Is this the focus of the page?
This document describes:
-  How to detect the dependency of `dockershim` in telemetry, logging, and security agents.
-  How to update telemetry, logging, security agents in your cluster.

kbhawkey commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,121 @@
+
+## Telemetry and security agents
+
+There are a few ways agents may run on kubernetes cluster. Agents may run on
```
s/kubernetes/Kubernetes/

kbhawkey commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,121 @@
+There are a few ways agents may run on kubernetes cluster. Agents may run on
+nodes directly or as DaemonSets.
+
+### Why do telemetry agents talk to Docker?
```
s/talk to/rely on/

kbhawkey commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,121 @@
+
+### SignalFx (Splunk) {#vendor-signalfx-splunk}
+
+The [SignalFx](https://www.splunk.com/en_us/investor-relations/acquisitions/signalfx.html)
```
Is this information available from the SignalFx/Splunk documentation?
Instead, readers should consult the telemetry and security agent documentation to learn how each application handles the `dockershim` migration.

kbhawkey commented:
Preview pages:
- https://deploy-preview-25787--kubernetes-io-master-staging.netlify.app/docs/tasks/administer-cluster/migrating-from-dockershim/
- https://deploy-preview-25787--kubernetes-io-master-staging.netlify.app/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/
- https://deploy-preview-25787--kubernetes-io-master-staging.netlify.app/docs/tasks/administer-cluster/migrating-from-dockershim/migrating-telemetry-and-security-agents/

SergeyKanzhelev commented:
/assign @ehashman 

@sftim @kbhawkey thank you for review. I think I addressed all the feedback. Can we move forward with it?

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/_index.md:
```diff
@@ -0,0 +1,21 @@
+Check out [container runtimes](/docs/setup/production-environment/container-runtimes/)
+section to know your options. Make sure to
+[report issues](https://github.com/kubernetes/kubernetes/issues) you encountered
+with the migration. So the issue can be fixed timely and your cluster would be
```
nit, not a merge blocker:
```suggestion
with the migration. So the issue can be fixed in at timely manner and your cluster would be
```

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,80 @@
+<!-- overview -->
+
+With Kubernetes 1.20 dockershim was deprecated. From the
+[Dockershim Deprecation FAQ](https://kubernetes.io/blog/2020/12/02/dockershim-faq/)
```
nit:
```suggestion
[Dockershim Deprecation FAQ](/blog/2020/12/02/dockershim-faq/)
```

sftim commented on content/en/docs/tasks/administer-cluster/migrating-from-dockershim/migrating-telemetry-and-security-agents.md:
```diff
@@ -0,0 +1,80 @@
+### Telemetry and security agent vendors
+
+We keep the work in progress version of migration instructions for various telemetry and security agent vendors
+in [Google doc](https://docs.google.com/document/d/1ZFi4uKit63ga5sxEiZblfb-c23lFhvy6RXVPikS8wf0/edit#).
```
Anyone can comment on this document. Is that a risk?

SergeyKanzhelev replied:
this is typical. I don't want to limit to kubernetes-dev group as security and telemetry vendors may not be a part of it

sftim reviewed (commented):
@kubernetes/sig-node-pr-reviews how does this look for technical accuracy?

Inline feedback is mainly minor details.

derekwaynecarr commented:
this looks good for sig-node technical accuracy.

/lgtm
/approve

dims commented:
/lgtm
/approve

sftim commented:
/unhold

Content guide concerns are addressed.

SergeyKanzhelev commented:
> My concern in https://github.com/kubernetes/website/pull/25787/files#r560586158 is about attracting vandalism. I don't know if that's much of a risk.

Google docs keeps history. I will go ahead and create a names snapshot of the doc and will keep them making snapshot for easy recovery from vandalism. I hope this addresses the concern.

SergeyKanzhelev commented:
@sftim this is ready to be merged.

SergeyKanzhelev commented:
Kindly ping @kbhawkey and @sftim

tengqm commented:
/approve
